cirq-core/cirq/ops/arithmetic_operation.py

```diff
@@ -241,8 +244,217 @@ def _apply_unitary_(self, args: 'cirq.ApplyUnitaryArgs'):
         return args.target_tensor
 
 
+TSelfGate = TypeVar('TSelfGate', bound='ArithmeticGate')
+
+
+class ArithmeticGate(Gate, metaclass=abc.ABCMeta):
+    """A helper gate for implementing reversible classical arithmetic.
+
+    Child classes must override the `registers`, `with_registers`, and `apply`
+    methods.
+
+    This class handles the details of ensuring that the scaling of implementing
+    the operation is O(2^n) instead of O(4^n) where n is the number of qubits
+    being acted on, by implementing an `_apply_unitary_` function in terms of
+    the registers and the apply function of the child class.
+
+    Examples:
+    ```
+
+        >>> class Add(cirq.ArithmeticGate):
+        ...     def __init__(self, target_register, input_register):
+        ...         self.target_register = target_register
+        ...         self.input_register = input_register
+        ...
+        ...     def registers(self):
+        ...         return self.target_register, self.input_register
+        ...
+        ...     def with_registers(self, *new_registers):
+        ...         return Add(*new_registers)
+        ...
+        ...     def apply(self, target_value, input_value):
+        ...         return target_value + input_value
```

Comment:
I don't think code written like this will pass mypy typing checks as it violates the Liskov Substitution Principle,  i.e. because the `apply` method on the `ArithmeticGate` base class has the signature `def apply(self, *register_values: int)`, in order for a derived class method to be compatible with the base class method, it must also be able to work with an arbitrary number of positional args (i.e. `register_values`). 

Here is an example of such a construct failing the mypy test because derived class method is not compatible with the base class method: 

https://mypy-play.net/?mypy=latest&python=3.10&gist=09e93ed48f7116c087e90da37cfd6a44

See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides for more details.

Reply:
I agree, but I tried it and mypy passes this example.

Reply:
Ah when I added the `int` type annotations it failed.